### PR TITLE
[install-expo-modules]: ensure swift_version build config is set

### LINKED ISSUE #30378: Unsupported Swift Version
### Minimal reproducible example

https://github.com/shuo-hiwintech/AwesomeProject

### What platform(s) does this occur on?

iOS

### Where did you reproduce the issue?

in a standalone app

### Summary

Dear Expo developers, Hello, I've run into a problem.

After installing Expo using the React Native CLI project, I get a warning: “Unsupported Swift Version” when I open the project through Xcode, but I don't get this warning when I open the project through Xcode without installing Expo.

image

Steps:
1. create a react native cli project, via "npx @react-native-community/cli@latest init AwesomeProject".
2. Install the Expo module via "npx install-expo-modules@latest".
3. Open the project via Xcode and get the Xcode warning.

Same question: https://github.com/expo/expo/issues/30337#issuecomment-2223156381

### Environment

expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.10.0 - /usr/local/bin/node
      Yarn: 3.6.4 - /usr/local/bin/yarn
      npm: 10.2.3 - /usr/local/bin/npm
      Watchman: 2024.04.15.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.4, iOS 17.4, macOS 14.4, tvOS 17.4, visionOS 1.1, watchOS 10.4
    IDEs:
      Android Studio: 2022.3 AI-223.8836.35.2231.11005911
      Xcode: 15.3/15E204a - /usr/bin/xcodebuild
    npmGlobalPackages:
      eas-cli: 10.0.2
      expo-cli: 6.3.0
    Expo Workflow: managed

### Expo Doctor Diagnostics

✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check for issues with metro config
✔ Check dependencies for packages that should not be installed directly
✖ Check for common project setup issues
✔ Check npm/ yarn versions
✔ Check for legacy global CLI installed locally
✔ Check that native modules do not use incompatible support packages
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that native modules use compatible support package versions for installed Expo SDK
✔ Check that packages match versions required by installed Expo SDK

### PR DESCRIPTION
**Pull request by vonovak**
# Why

closes #30378. The error there (Unsupported Swift Version) is a bit misleading because the code is written with Swift 5 syntax. The real issue is that `SWIFT_VERSION` build config is not set.

# How

We go over the `XCBuildConfiguration` entries and set `SWIFT_VERSION` if it's not present. Given that the xcode target in question isn't really under our control, I opted for this approach.

# Test Plan

tested manually, plus added a test

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/install-expo-modules/src/plugins/ios/withSwiftVersion.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { type ConfigPlugin, withXcodeProject } from '@expo/config-plugins';
+import type { XcodeProject } from 'xcode';
+
+export function setSwiftVersionIfNotPresent(
+  swiftVersionToSet: string,
```
Comment (vonovak) [inline packages/install-expo-modules/src/plugins/ios/withSwiftVersion.ts:5]: I wonder if I should be providing the `swiftVersionToSet` or detecting it from other targets?

File: packages/install-expo-modules/tsconfig.json [hunk 2]
```diff
@@ -2,7 +2,10 @@
   "extends": "../../tsconfig.node",
   "compilerOptions": {
     "outDir": "build",
-    "rootDir": "src"
+    "rootDir": "src",
+    "typeRoots": [
+      "../@expo/cli/ts-declarations",
```
Comment (vonovak) [inline packages/install-expo-modules/tsconfig.json:7]: provides typings for `xcode` package

File: packages/install-expo-modules/package.json [hunk 3]
```diff
@@ -16,7 +16,7 @@
     "build"
   ],
   "scripts": {
-    "build": "ncc build ./src/index.ts -o build/",
+    "build": "ncc build --external simple-plist ./src/index.ts -o build/",
```
Comment (vonovak) [inline packages/install-expo-modules/package.json:19]: in `packages/install-expo-modules/build/index.js` most requires have this form: `__nccwpck_require__(92740)`. But `simple-plist` keeps referring to relative paths: `require("./parse")`. And that fails.

Comment (byCedric) [reply packages/install-expo-modules/package.json:19]: I'm slightly concerned by this, as `build:prod` should bundle everything. If somehow `simple-plist` doesn't bundle in `build`, I think it won't bundle in `build:prod` too.

File: packages/install-expo-modules/src/plugins/ios/__tests__/withSwiftVersion-test.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+const fixturesPath = path.resolve(__dirname, 'fixtures');
+
+it(
+  "given targets in pbxproj that don't have `SWIFT_VERSION` build config specified, " +
+    'setSwiftVersionIfNotPresent() sets it',
+  async () => {
+    const project = xcode.project(
+      path.join(fixturesPath, 'TestTargetsDontHaveSwiftVersion.pbxproj')
+    );
+    project.parseSync();
```
Comment (byCedric) [inline packages/install-expo-modules/src/plugins/ios/__tests__/withSwiftVersion-test.ts:16]: I think we can use the `IOSConfig.XcodeUtils.getPbxproj` helper from `@expo/config-plugins` here, it should do the same without `install-expo-modules` having to depend on the `xcode` package.

Something like:

```
import {} from '@expo/config-plugins';

IOSConfig.XcodeUtils.getPbxproj(path.join(fixturePath, 'TestTargetsDontHaveSwitftVersion.pbxproj'))
```

But that might require the fixture to have an `ios` folder, e.g. `ios/TestTargetsDontHaveSwiftVersion.pbxproj`.

**Review by byCedric — COMMENTED** [review]
I think @Kudo might be better here to put the stamp of approval here 😄  Looks good though!

**Review by Kudo — CHANGES_REQUESTED** [review]
thanks for having the pr. i encountered the same simple-plist issue previously and that's why i replaced it with xcparse (now called `@bacons/xcode`) in https://github.com/expo/expo-cli/pull/4554. could you try to use xcparse in this pr? i guess we will ultimately to use `@bacons/xcode` in all config-plugins at some point.

File: packages/install-expo-modules/src/plugins/ios/__tests__/withSwiftVersion-test.ts (hunk 4 above)
Comment (vonovak) [reply packages/install-expo-modules/src/plugins/ios/__tests__/withSwiftVersion-test.ts]: ended up using `xcparse` which is the package that `withXCParseXcodeProjectBaseMod` is built on top of

**Review by Kudo — APPROVED** [review]
having xcparse is much clear. thanks again for updating and the pr!

File: packages/install-expo-modules/src/plugins/ios/withSwiftVersion.ts (hunk 1 above)
Comment (vonovak) [reply packages/install-expo-modules/src/plugins/ios/withSwiftVersion.ts:7]: right now it's going to be set to 5.0

**Pull request merged**